dwave/optimization/src/nodes/indexing.cpp

```diff
@@ -248,22 +248,47 @@ struct AdvancedIndexingNode::IndexParser_ {
             this->ndim += std::holds_alternative<Slice>(index);
 
             if (std::holds_alternative<ArrayNode*>(index)) {
-                ssize_t a_ndim = std::get<ArrayNode*>(index)->ndim();
+                ArrayNode* indexing_array_ptr = std::get<ArrayNode*>(index);
+
+                // Test whether we can index with the given array. This is probably a bit
+                // over-strict and may need to be loosened over time.
+                if (!indexing_array_ptr->integral()) {
+                    throw std::invalid_argument("cannot index an array with a non-integral array");
+                }
+                if (indexing_array_ptr->min() < 0) {
+                    throw std::invalid_argument("indexing array cannot allow negative values");
+                }
+                if (idx == 0 && array_ptr->shape()[0] < 0) {
+                    // 100 is a magic number
```

Comment:
This is the maximum graph depth to search for "tightening" the size info?

Reply:
More like loosening.

Reply:
Yes